### PR TITLE
Implement endpoints controller for service without selector

### LINKED ISSUE #1694: Service without selector needs endpoints for internal service
**What would you like to be added**:
Support globalnet for service without selector.

**Why is this needed**:
With https://github.com/submariner-io/submariner/pull/1643, iptables rule to forward to kube-proxy chain for service is removed, instead forwarding to globalIngressIP to endpoints is offloaded to externalIP feature of the Service, by creating internal service. Endpoints of the service __with__ selector can be synced between original and internal services by sharing labels, but Endpoints of the service __without__ selector needs to be handled in a different way.

### PR DESCRIPTION
**Pull request by mkimuram**
This PR implement endpoints controller for service without selector.

Dependency:
- [x] https://github.com/submariner-io/submariner-operator/pull/1845

Fixes: https://github.com/submariner-io/submariner/issues/1694

### CODE REVIEW

**Comment by mkimuram** [comment]
@sridhargaddam @nerdalert @anfredette 

We might need to discuss how it should be implemented, first. However, to make it easier to discuss further, I've implemented it to just sync original endpoints to cloned endpoints if the original endpoints has "endpoints.submariner.io/exported: true" label. By doing so, the sync logic can be completely separated from other globalnet's behaviors, like if the endpoints are for service without selector or exported or if the service for the endpoints  even exists.

Note that we would need to add RBAC rule to allow globalnet controller to create endpoints.

File: pkg/globalnet/controllers/endpoints_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+var _ = Describe("Endpoints controller", func() {
+	t := newEndpointsControllerTestDriver()
+
+	var (
+		epName string
+		ep     *corev1.Endpoints
+	)
+
+	When("an endpoints without label is created", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/endpoints_controller_test.go]:
```suggestion
	When("an Endpoints resource without the exported label is created", func() {
```

File: pkg/globalnet/controllers/endpoints_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			t.createEndpoints(newEndpoints(epName, map[string]string{}))
+			t.awaitEndpoints(epName)
+		})
+
+		It("should not create a cloned endpoints", func() {
+			t.awaitNoEndpoints(controllers.GetInternalSvcName(epName))
+		})
+	})
+
+	When("an endpoints with label is created", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/endpoints_controller_test.go]:
```suggestion
	When("an Endpoints resource with the exported label is created", func() {
```

File: pkg/globalnet/constants/constants.go [hunk 3]
```diff
@@ -31,5 +31,7 @@ const (
 	SmGlobalnetEgressChainForNamespace       = "SM-GN-EGRESS-NS"
 	SmGlobalnetEgressChainForCluster         = "SM-GN-EGRESS-CLUSTER"
 
-	SmGlobalIP = "submariner.io/globalIp"
+	SmGlobalIP         = "submariner.io/globalIp"
```
Comment (tpantelis) [inline pkg/globalnet/constants/constants.go]: We can drop the _Sm_ prefix.

File: pkg/globalnet/controllers/endpoints_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		BeforeEach(func() {
+			epName = "epWithLabel"
+			ep = t.createEndpoints(newEndpoints(epName, map[string]string{"endpoints.submariner.io/exported": "true"}))
```
Comment (tpantelis) [inline pkg/globalnet/controllers/endpoints_controller_test.go]:
```suggestion
			ep = t.createEndpoints(newEndpoints(epName, map[string]string{constants.ExportedEndpoint: "true"}))
```

File: pkg/globalnet/constants/constants.go [hunk 5]
```diff
@@ -31,5 +31,7 @@ const (
...
+	SmExportedEndpoint = "endpoints.submariner.io/exported"
+	SmLabelTrue        = "true"
```
Comment (tpantelis) [inline pkg/globalnet/constants/constants.go]: A bit of a nit but I think it's overkill to create a constant for "_true_".  But if we do I think we should be consistent and define it in admiral and use it everywhere.

File: pkg/globalnet/controllers/endpoints_controller.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:            "Endpoints syncer",
+		ResourceType:    &corev1.Endpoints{},
+		SourceClient:    config.SourceClient,
+		SourceNamespace: corev1.NamespaceAll,
+		RestMapper:      config.RestMapper,
+		Federator:       federate.NewCreateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Transform:       controller.process,
+	})
+
```
Comment (tpantelis) [inline pkg/globalnet/controllers/endpoints_controller.go:69]: To be complete and consistent with other controllers, we probably should reconcile possible internal Endpoints whose exported Endpoints were deleted during downtime.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/globalnet/controllers/endpoints_controller.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	if val, ok := ep.Labels[constants.SmExportedEndpoint]; !ok || val != constants.SmLabelTrue {
+		return nil, false
+	}
+
+	switch op {
+	case syncer.Create:
+		return c.onCreate(ep)
+	case syncer.Delete:
+		return c.onDelete(ep)
+	case syncer.Update:
```
Comment (sridhargaddam) [inline pkg/globalnet/controllers/endpoints_controller.go]: Currently, we are skipping the Update event. What if the user modifies the ip/port of the original endpoint?

File: pkg/globalnet/controllers/endpoints_controller.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func (c *endpointsController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	ep := from.(*corev1.Endpoints)
+
+	// Skip endpoints without "endpoints.submariner.io/exported: true" label
```
Comment (sridhargaddam) [inline pkg/globalnet/controllers/endpoints_controller.go]: This logic works with an assumption that user will create original endpoints with a specific label `endpoints.submariner.io/exported`. But its very likely that user might miss adding this label to the original endpoint as they will just create regular service without selectors and an associated endpoint. IMHO we should make that use-case work out of the box without expecting this label from the user.

**Review by sridhargaddam — CHANGES_REQUESTED** [review]
Thanks for the PR @mkimuram

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 8 above)
Comment (vthapar) [reply pkg/globalnet/controllers/endpoints_controller.go]: We need some means to know if endpoint is created by user for Exported Service or not. If no label, then we have to check if ServiceExport is created or not. Also, we have to handle case where ServiceExport is created after Endpoint is created.

Requirement to have label, though minor, is not without precedent in kubernetes. We can handle Update where one may add label after creation when they decide to export service.

Still, we can't add controller for *all* endpoints as there will be way too many endpoints in a cluster. Note lighthouse ads label for endpointslices created by it. We can easily add code in `subctl` export service to add label for external services.

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 7 above)
Comment (vthapar) [reply pkg/globalnet/controllers/endpoints_controller.go]: We should also handle the case where user may add the label later. Possible for scenarios where external service already existed but now chose to export it.

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 8 above)
Comment (sridhargaddam) [reply pkg/globalnet/controllers/endpoints_controller.go]: > Still, we can't add controller for _all_ endpoints as there will be way too many endpoints in a cluster. Note lighthouse ads label for endpointslices created by it. We can easily add code in `subctl` export service to add label for external services.

We don't need a controller for all endpoints. We can create a controller/listener with a label selector (just like how we are creating the pod listeners) where the selector will match the name of the original exported service.

**Comment by mkimuram** [comment]
@tpantelis @sridhargaddam @vthapar

Thank you for your comments. I will update the PR as suggested.

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 8 above)
Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller.go]: @sridhargaddam 

I've changed the codes to handle only Endpoints for exported Services. Current results of unit tests shows that it works well regardless of the order of exporting Services and creating Endpoints, however, it doesn't seems to work if original Endpoints resource is updated after cloned Endpoints are created by the controller (in this case, the cloned Endpoints is also expected to be updated to have the same Subsets, but it doesn't seems to be updated).

Could you check if I'm doing something wrong?
(Are there any requirements for admiral to actually update the resource? I assume that just returning the expected resource in `Transform` function makes the resource updated).

**Comment by anfredette** [comment]
@mkimuram, as an FYI, I tested your original PR on some KIND clusters today, and I couldn't get it to work, though it's possible I was doing something wrong.  I'll follow up with you on it tomorrow.  I can give your new one a try tomorrow as well.

**Comment by mkimuram** [comment]
@anfredette

Thank you for your testing.
You can test this PR like below way (although, it still has an [issue](https://github.com/submariner-io/submariner/pull/1695#discussion_r807450882)).

<details>

1. Run e2e test for deploy the test environment (test itself should fail)
```
make e2e using=external-net,globalnet
```

2. Update submariner-globalnet clusterrole and recreate globalnet pod (https://github.com/submariner-io/submariner-operator/pull/1845 will fix this)

```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
kubectl get clusterrole submariner-globalnet -o json \
  | jq '.rules += [{"apiGroups":[""],"resources":["endpoints"],"verbs":["create","get","list","watch","update","delete"]}]' \
  | kubectl apply -f -
kubectl delete pod -l "app=submariner-globalnet" -n submariner-operator

export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
kubectl get clusterrole submariner-globalnet -o json \
  | jq '.rules += [{"apiGroups":[""],"resources":["endpoints"],"verbs":["create","get","list","watch","update","delete"]}]' \
  | kubectl apply -f -
kubectl delete pod -l "app=submariner-globalnet" -n submariner-operator
```

3. Create Service without selector and Endpoints (`Endpoints.subsets.addresses.ip` should point to test-vm docker container's IP)

```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: test-vm
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
EOF

cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: 172.117.0.4
    ports:
      - port: 80
EOF

subctl export service -n default test-vm
```

4. Confirm that internal Endpoints is created by controller and globalingressip is allocated

```
kubectl get endpoints
NAME                                          ENDPOINTS         AGE
kubernetes                                    172.18.0.7:6443   29m
submariner-4cl54k5nvyvsg3p65n5la4lbtg5scllc   172.117.0.4:80    6m51s
test-vm                                       172.117.0.4:80    6m56s

kubectl get globalingressip test-vm -o yaml | grep allocatedIP
  allocatedIP: 242.254.1.253
```

5. Confirm that test-vm can be accessed from pod in cluster2

```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
kubectl -n default run tmp-shell --rm -i --tty --image quay.io/submariner/nettest -- bash
curl 242.254.1.253
```

</details>

**Comment by anfredette** [comment]
@mkimuram, FYI, today I just checked out your pr, and ran the following:
`make deploy using=lighthouse DEPLOY_ARGS='--globalnet --cable_driver vxlan'`

Without doing anything else, it failed the basic make deploy check.  [Details here](https://gist.github.com/anfredette/d8be2639ba2ec8dc8e1359d0474401e3).

Without your pr, the above make deploy ran fine and I was able to do my manual testing for service without selector.  I'll give your test approach a try also.

**Review by anfredette — CHANGES_REQUESTED** [review]
See comments made on the "Conversation" tab.

**Comment by mkimuram** [comment]
@anfredette 

Could you try `make deploy using=external-net,globalnet,lighthouse,vxlan` instead of passing similar args to `DEPLOY_ARGS` directly? In Makefile, at least `external-net` updates other than `DEPLOY_ARGS`, so it may fail due to this difference (I didn't test with `lighthouse` and `vxlan`, but I hope that it should work).

**Comment by anfredette** [comment]
Will do. I'm trying your approach first. Then I'll come back and give this
a try.

**Comment by anfredette** [comment]
> kubectl get clusterrole submariner-globalnet -o json \
>   | jq '.rules += [{"apiGroups":[""],"resources":["endpoints"],"verbs":["create","get","list","watch","update","delete"]}]' \
>   | kubectl apply -f -

If I checkout and build with https://github.com/submariner-io/submariner-operator/pull/1845, do I still need to run the above command?

**Comment by mkimuram** [comment]
Please check if clusterrole submariner-globalnet has ["create","get","list","watch","update","delete"] right for apiGroups [""]'s "endpoints" resource. If not, you need to. Just for testing, it would be safe to run the command any time, because it just adds above, even if it already exists.

**Comment by mkimuram** [comment]
> If I checkout and build with https://github.com/submariner-io/submariner-operator/pull/1845, do I still need to run the above command?

The deploy script for e2e uses the subctl command inside the dappper, so you would need to do it even if you build subctl by yourself (unless you replace the subctl inside dapper). If you manually deploy your environment by using the subctl command, it would be work well without the command.

**Comment by anfredette** [comment]
Got it.

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 8 above)
Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller.go]: @tpantelis @sridhargaddam 

I noticed there is `NewCreateOrUpdateFederator`, not `NewCreateFederator`. Is my understanding right that we need to use `NewCreateOrUpdateFederator` if updating resource is required?
If so, what needs to be passed as `localClusterID` and `keepMetadataField`?
(At a glance, just passing "" and nothing seems working, but I need to ensure that it is the right way).

Comment (tpantelis) [reply pkg/globalnet/controllers/endpoints_controller.go]: > @tpantelis @sridhargaddam
> 
> I noticed there is `NewCreateOrUpdateFederator`, not `NewCreateFederator`. Is my understanding right that we need to use `NewCreateOrUpdateFederator` if updating resource is required? 

Yes

If so, what needs to be passed as `localClusterID` and `keepMetadataField`? (At a glance, just passing "" and nothing seems working, but I need to ensure that it is the right way).

`localClusterID` is used for 2-way syncing with the broker - that's not relevant here so pass "". You also don't need any `keepMetadataField`.

Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller.go]: @tpantelis Thank you for your quick answer. I will update the code as so.

**Review by anfredette — COMMENTED** [review]
I re-ran the test following Masaki's instructions above, and it does work. 

I then tested with lighthouse by running `make e2e using=lighthouse,external-net,globalnet` to set up the clusters. I also used [this script](https://github.com/anfredette/kube-scripts/blob/main/create-service-wo-selector.sh) that I used for testing https://github.com/submariner-io/lighthouse/pull/684 to set up the service without selector. This script uses a regular pod as an endpoint instead of an external VM, but I think the Submariner flows are the same in either case. That test worked as well and I confirmed that the services names could be resolved from cluster2. I.e., `curl nginx-svc.default.svc.clusterset.local` worked in this example.

One observation I have is that the code creates an internal service, called submariner-uvhtheozeyskii6xhvh5dxgtbdktupw3, as well as an internal endpoint with the same name, and assigns the external IP address to the internal service.

```
$ kubectl get services
NAME                                          TYPE        CLUSTER-IP     EXTERNAL-IP     PORT(S)   AGE
kubernetes                                    ClusterIP   100.0.0.1      <none>          443/TCP   176m
nginx-svc                                     ClusterIP   100.0.138.96   <none>          80/TCP    161m
submariner-uvhtheozeyskii6xhvh5dxgtbdktupw3   ClusterIP   100.0.11.170   242.254.1.253   80/TCP    161m
$ kubectl get endpoints
NAME                                          ENDPOINTS         AGE
kubernetes                                    172.18.0.4:6443   176m
nginx-svc                                     10.0.128.5:80     161m
submariner-uvhtheozeyskii6xhvh5dxgtbdktupw3   10.0.128.5:80     161m
```

This may be a newbie misunderstanding, but is the internal service necessary? For my manual workaround documented in https://github.com/submariner-io/submariner/issues/1693, I only needed to manually create the internal endpoint to get the datapath to work. If we had 10 endpoints, would we need to create 10 extra services? Would they all get different external IP addresses or the same? Furthermore, why does globalnet need to care about all the endpoints? Wouldn't the local service loadbalancer be responsible for load balancing the sessions across the endpoints?

We can discuss this more during the non-cluster connectivity meeting tomorrow morning. Thanks!

**Comment by mkimuram** [comment]
@anfredette In short, after #1643,
- For non-exported Service, globalnet controller do nothing (as it was),
- For exported Service with selector, globalnet controller creates internal Service, and Endpoints for it is created by k8s,
- For exported Service without selector, globalnet controller creates internal Service, but Endpoints for it isn't created by k8s.

So, this PR is to add codes to automatically create Endpoints for internal Service, if Service is Service without selector (as you documented to manually do in #1693). I know that this explanation won't be enough, so let's discuss in the meeting.

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 8 above)
Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller.go]: @sridhargaddam  I've updated the PR and it now can handle update. PTAL
 
The last commit is needed for deleting cloned Endpoints if original Endpoints is deleted while controller stops, but it avoids controller become running unless RBAC isn't granted (as handled in submariner-io/submariner-operator#1845 , so it needs to be merged, first).

File: pkg/globalnet/controllers/endpoints_controller_test.go (hunk 1 above)
Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller_test.go]: Test codes are moved to pkg/globalnet/controllers/service_export_controller_test.go and changed in this way.

File: pkg/globalnet/constants/constants.go (hunk 3 above)
Comment (mkimuram) [reply pkg/globalnet/constants/constants.go]: Removed prefix.

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 6 above)
Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller.go:69]: Add codes at `ensureClonedHasOriginal` in pkg/globalnet/controllers/service_export_endpoints_controllers.go

File: pkg/globalnet/controllers/endpoints_controller.go (hunk 7 above)
Comment (mkimuram) [reply pkg/globalnet/controllers/endpoints_controller.go]: Add codes to handle update and it can handle exporting service before/after endpoints creation.

**Comment by mkimuram** [comment]
@tpantelis @sridhargaddam @vthapar

I've squashed the commit and it is now ready to be reviewed. PTAL

Note that the second commit is needed because ExternalIP feature of k8s doesn't seem to forward packets from pod on the same cluster when accessed via ExternalIP, therefore this access can't be reachable. However, it won't need to be covered, because there is no use case for it (This cluster is just to connect cluster and external, so no user pod will be deployed). 
Cleanup for E2E tests will be done in https://github.com/submariner-io/submariner/pull/1582.

@dfarrell07

Now that submariner-operator#1845 is merged, this PR works well in my environment (without any manual interaction).
However, it seems that CI is skipping the test. Could you check what is happening (is it just taking time)?

**Comment by dfarrell07** [comment]
@mkimuram It just needed the `ready-to-test` label, which I've added. It's added automatically by the bot when there are two +1 reviews, or it's fine to add it manually before then if you need to see a result like in this case.

**Review by tpantelis — APPROVED** [review]

**Review by anfredette — APPROVED** [review]
LGTM

I've also tested this code successfully using various versions of the following scripts:
https://github.com/anfredette/kube-scripts/blob/main/create-service-wo-selector.sh
https://github.com/anfredette/kube-scripts/blob/main/create-multi-service-wo-selector.sh

**Review by sridhargaddam — APPROVED** [review]
Awesome work @mkimuram

**Comment by anfredette** [comment]
FYI, I rebased this pr since @mkimuram is out this week.

**Comment by sridhargaddam** [comment]
Thank you @anfredette

**Pull request merged**